Bringup Java 11 b04 - implement j.l.String.valueOfCodePoint()

JasonFengJ9 opened the pull request:
Bringup `Java 11 b04` - implement `j.l.String.valueOfCodePoint()`

Implemented `j.l.String.valueOfCodePoint()`, manually verified that the output of `String.valueOfCodePoint(codePoint).codePointAt(0)` equals `codePoint` for all Unicode code point.

closes: #1439

Reviewer @DanHeidinga @fjeremic 

Signed-off-by: Jason Feng <fengj@ca.ibm.com>

DanHeidinga commented:
Change looks reasonable to me.  @fjeremic Any comments on the string compression part?

pshipton commented:
Rather than manual verification, please add the automated test.

Did you manually test with string compression enabled?

JasonFengJ9 commented:
Manually verified with both command line options `-XX:+CompactStrings` and `-XX:-CompactStrings`.

Will work on automated tests tomorrow.

fjeremic commented on jcl/src/java.base/share/classes/java/lang/String.java:
```diff
@@ -5354,4 +5354,41 @@ public IntStream codePoints() {
+			if (enableCompression) {
+				string = new String(compressedAsciiTable[codePoint], 0, 1, true);
+			} else {
+				string = new String(decompressedAsciiTable[codePoint], 0, 1, false);
```
Would it be possible to use:

https://github.com/JasonFengJ9/openj9/blob/c864cc795b78b3b1c573e8bc5684c7f9f413c6af/jcl/src/java.base/share/classes/java/lang/String.java#L402

Constructor here instead? The one currently being used will have to perform many of the checks again. Similarly for all other assignments to `string` in this method.

JasonFengJ9 replied:
`String(byteArray, coder)` works as well, will update the PR accordingly.

fjeremic commented on jcl/src/java.base/share/classes/java/lang/String.java:
```diff
@@ -5354,4 +5354,41 @@ public IntStream codePoints() {
+		String string;
+		if ((codePoint < Character.MIN_CODE_POINT) || (codePoint > Character.MAX_CODE_POINT)) {
+			/*[MSG "K0800", "Invalid Unicode code point"]*/
+			throw new IllegalArgumentException(com.ibm.oti.util.Msg.getString("K0800"));	//$NON-NLS-1$
```
Would it be possible to put case as the last `else` case as it occurs least often. The JIT will have to do less work when block shuffling and if it every got it wrong we wouldn't pay the performance penalty.

DanHeidinga replied:
Wouldn't the next check for `<= 255` need to expand then?  It would also need to validate codePoint > MIN_CODE_POINT.

JasonFengJ9 replied:
Will following checking help JIT performance?
```
if ((codePoint < Character.MIN_CODE_POINT)  {
  // error handling 
} else if (codePoint <= 255) {
} else if (codePoint < Character.MIN_SUPPLEMENTARY_CODE_POINT) {
} else if (codePoint <= Character.MAX_CODE_POINT) {
} else {
  // error handling
}
```
This saves the less often check `(codePoint > Character.MAX_CODE_POINT)` to the last.

JasonFengJ9 commented:
@fjeremic the PR is ready for another review with following updates:
1.  Uses constructor `String(byteArray, coder)`;
2. Modified the order to save less often check `(codePoint > Character.MAX_CODE_POINT)` last;
3. Added tests. Note: this test suite is for `Java 11` hence will run when `Java 11` are introduced to the build jobs. Manually verified that it passes against `Java 11` raw builds with modified command line options.

@llxia please arrange a test team review.

Additional note: three project files failed Copyright check however this kind of files don't have copyright header either such as `Java[10|9|8]andUp`.

DanHeidinga commented on jcl/src/java.base/share/classes/java/lang/String.java:
```diff
@@ -5354,4 +5354,45 @@ public IntStream codePoints() {
+		// Save less often check (codePoint > Character.MAX_CODE_POINT) last
+		if (codePoint < Character.MIN_CODE_POINT) {
+			/*[MSG "K0800", "Invalid Unicode code point"]*/
+			throw new IllegalArgumentException(com.ibm.oti.util.Msg.getString("K0800"));	//$NON-NLS-1$
```
Can you include the string or the the code point in the exception?  It will make it much easier to debug one of these issues.

JasonFengJ9 replied:
Yup, now an exception looks like:
```
FAILED: testValueOfCodePoint
java.lang.AssertionError: Unicode code point mismatch at 0 expected [1] but found [0]
	at org.testng.Assert.fail(Assert.java:96)
	at org.testng.Assert.failNotEquals(Assert.java:776)
	at org.testng.Assert.assertEqualsImpl(Assert.java:137)
	at org.testng.Assert.assertEquals(Assert.java:118)
	at org.testng.Assert.assertEquals(Assert.java:652)
	at org.openj9.test.java_lang.Test_String.testValueOfCodePoint(Test_String.java:49)
```

DanHeidinga commented on test/Java11andUp/src/org/openj9/test/java_lang/Test_String.java:
```diff
@@ -0,0 +1,73 @@
+				try {
+					String str = (String) method.invoke(null, i);
+					Assert.assertEquals(i, str.codePointAt(0), "Unicode code point mismatch at " + i);
+				} catch (IllegalAccessException | IllegalArgumentException | InvocationTargetException e) {
```
Test cases shouldn't catch unexpected exceptions.  Let the test code throw them and the test harness will (usually) print the relevant stacktrace.

JasonFengJ9 replied:
Fixed, removed catching of unexpected exceptions.

llxia commented:
@renfeiw , could you review the test change? Thanks.

fjeremic commented on jcl/src/java.base/share/classes/java/lang/String.java:
```diff
@@ -5354,4 +5354,45 @@ public IntStream codePoints() {
+	static String valueOfCodePoint(int codePoint) {
+		String string;
+		// Save less often check (codePoint > Character.MAX_CODE_POINT) last
+		if (codePoint < Character.MIN_CODE_POINT) {
```
I would vouch to say this is also an uncommon path and should be moved with the `else` path. Do we agree?

DanHeidinga replied:
We need to check that before checking `} else if (codePoint <= 255) {` or we need to move it into the check for `<= 255`.  Either way, its got to be first.

fjeremic replied:
Sure, but merging the two checks has the benefit of smaller footprint and less code to jump over to get to the `0 <= codePoint <= 255` case.

DanHeidinga replied:
I strongly dislike the duplication of the second form.  If we can't optimize the first form reasonably then we're going to have problems with most user code.

JasonFengJ9 replied:
It seems you are suggesting code like following:
```
if (codePoint >= Character.MIN_CODE_POINT && codePoint <= 255) {
  ...
} else if (codePoint >= Character.MIN_CODE_POINT && codePoint < Character.MIN_SUPPLEMENTARY_CODE_POINT) {
  ...
} else if (codePoint >= Character.MIN_CODE_POINT && codePoint <= Character.MAX_CODE_POINT) {
  ...
} else {
  error handling
}
```
This introduces extra code and confusion (later reader need lots of explanation about the rational of this code logic). I don't think any performance gain here worth the potential issue later. 
Now I feel this code should go back to original structure, i.e.,
```
if ((codePoint < Character.MIN_CODE_POINT) || (codePoint > Character.MAX_CODE_POINT)) {
  error handling
} else if (codePoint <= 255) {
  ...
} else if (codePoint < Character.MIN_SUPPLEMENTARY_CODE_POINT) {
  ...
} else {
  ...
}
```
It is clear and easy to understand. Can we modify it to improve performance later if this becomes a hot spot?

fjeremic replied:
Yes, I hadn't thought about the extra checks introduced in my original comment. I'm fine either keeping as is or reverting back to your latest suggestion @JasonFengJ9. Apologies for the churn!

JasonFengJ9 replied:
No problem, updated PR as latest suggestion above.

renfeiw commented:
To enable SE110, we need to make an update in testKitGen. (https://github.com/eclipse/openj9/pull/1495)

JasonFengJ9 commented on test/Java11andUp/build.xml:
```diff
@@ -0,0 +1,88 @@
+	
+	<target name="build" >
+		<if>
+			<equals arg1="${JAVA_VERSION}" arg2="SE110"/>
```
Agreed, changed the condition to `not ^SE(8|9|10)0$$`.

JasonFengJ9 commented:
The PR has been updated to address review comments.
Though `Java 11` hasn't been built and run in any test job yet, I manually verified that `Java11andUp` test project can be built, run and detect problem as expected.
Please review again.

DanHeidinga commented on jcl/src/java.base/share/classes/java/lang/String.java:
```diff
@@ -5354,4 +5354,42 @@ public IntStream codePoints() {
+	 */
+	static String valueOfCodePoint(int codePoint) {
+		String string;
+		// Save less often check (codePoint > Character.MAX_CODE_POINT) last
```
Comment is no longer accurate

JasonFengJ9 replied:
The comment in question has been removed.

DanHeidinga commented on test/Java11andUp/src/org/openj9/test/java_lang/Test_String.java:
```diff
@@ -0,0 +1,60 @@
+		// Character.MIN_CODE_POINT - 1 & Character.MAX_CODE_POINT + 1
+		try {
+			method.invoke(null, Character.MIN_CODE_POINT - 1);
+			method.invoke(null, Character.MAX_CODE_POINT + 1);
```
The second method.invoke will never run as the first show throw.  We need this to be two tests or to have it written like:
```
try {
method.invoke(null, Character.MIN_CODE_POINT - 1);
fail()
} catch (IllegalArgumentException | InvocationTargetException e) {} 
try {
method.invoke(null, Character.MIN_CODE_POINT - 2);
fail()
} catch (IllegalArgumentException | InvocationTargetException e) {} 
```

JasonFengJ9 replied:
Updated the test with suggested format above.
Please have another look.

DanHeidinga commented on test/Java11andUp/src/org/openj9/test/java_lang/Test_String.java:
```diff
@@ -0,0 +1,64 @@
+		}
+		// Verify that IllegalArgumentException is thrown for Character.MIN_CODE_POINT - 1.
+		try {
+			method.invoke(null, Character.MIN_CODE_POINT - 1);
```
This needs an explicit `Assert.fail()` call or the method could succeed and pass the test when we know it should throw the exception.

JasonFengJ9 replied:
Agreed, added `Assert.fail(message)`.

DanHeidinga commented on test/Java11andUp/src/org/openj9/test/java_lang/Test_String.java:
```diff
@@ -0,0 +1,64 @@
+		}
+		// Verify that IllegalArgumentException is thrown for Character.MAX_CODE_POINT + 1.
+		try {
+			method.invoke(null, Character.MAX_CODE_POINT + 1);
```
This needs an explicit `Assert.fail()` call or the method could succeed and pass the test when we know it should throw the exception.

JasonFengJ9 replied:
Fixed as well. Please have another look, thanks.

DanHeidinga commented:
Jenkins test sanity plinux

DanHeidinga commented:
Running one builds worth of sanity to ensure that this is correctly ignored by older Java levels.
